### ISSUE TITLE
gpuinfo reports wrong workitem dimensions for 32-bit architectures

### ISSUE THREAD

**Issue by doe300** [hughperkins/EasyCL#24]
The value returned by `clGetDeviceInfo` with the parameter `CL_DEVICE_MAX_WORK_ITEM_SIZES` is of type `size_t[]` and is read by `easycl::printDeviceInfoArray()` as `cl_ulong[]`. 
Since for 32-bit systems, `size_t` is `unsigned int`, not `unsigned long`, the value read is wrong.

**Comment by hughperkins** [comment]
Ok. Thoughts on how to fix this?

**Comment by doe300** [comment]
If all device-info queries which use `easycl::printDeviceInfoArray()` have the type `size_t[]`, you could just rewrite this method to use a size_t array.

Otherwise, copy `easycl::printDeviceInfoArray()` and write a version for arrays of type size_t. Or make it a template-function with specializations for both `cl_ulong`and `size_t` and use the correct one.

**Comment by hughperkins** [comment]
First option sounds tempting, but it's not very general, is that a fair impression?

Second option sounds plausible, and easy-ish. Would need to keep the old function in, for backwards compatibility.

One variant on the second is, looks like `clGetDeviceInfo` can tell us the size in bytes of the data being returned? https://www.khronos.org/registry/OpenCL/sdk/1.0/docs/man/xhtml/clGetDeviceInfo.html

**Comment by doe300** [comment]
Yeah, 1 is the more simple one, but will introduce other errors, if for some query `cl_ulong[]` is returned. Which, as far as I can tell, is never used as a return-type for a device-query in OpenCL 2.1, but may be used by extensions or future versions.

As to getting the byte-size of the type returned: The type-size can be calculated, and then you could use formatted printing (with either `%u`or `%ul`) depending on the byte-size of the type.

**Comment by hughperkins** [comment]
> As to getting the byte-size of the type returned: The type-size can be calculated, and then you could use formatted printing (with either %uor %ul) depending on the byte-size of the type.

Yes. Well, at least have a `case` switch, with the `size_t` and `cl_long` types for now, and an exception for other types.
